### PR TITLE
refactor(util): add gen_effect_id TypeScript version

### PR DESCRIPTION
**Pull request by MaikoTan**

### CODE REVIEW

File: util/gen_effect_id.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+  const foundNames = new Set();
+
+  const map: { [s: string]: string } = {};
+  for (const [id, effect] of Object.entries(table)) {
+    const rawName = effect['Name'];
+    if (!rawName)
+      continue;
+    const name = cleanName(rawName);
+
+    if (rawName && rawName in knownMapping) {
```
Comment (quisquous) [inline util/gen_effect_id.ts:80]:
```suggestion
    if (rawName in knownMapping) {
```

File: util/gen_effect_id.ts [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+    foundNames.add(name);
+    map[name] = id;
+  }
+
+  // Make sure everything specified in known_mapping was found in the above loop.
+  for (const entry of Object.entries(knownMapping)) {
+    const name = cleanName(entry[0]);
+    if (name && !(name in foundNames))
+      printError('missing', name, knownMapping, entry[0]);
```
Comment (quisquous) [inline util/gen_effect_id.ts:106]:
```suggestion
  for (const rawName of Object.keys(knownMapping)) {
    const name = cleanName(rawName);
    if (name && !(name in foundNames))
      printError('missing', name, knownMapping, rawName);
```

File: util/gen_effect_id.ts [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  }
+
+  // Add custom effect name for necessary duplicates.
+  for (const [rawName, id] of Object.entries(customMapping))
+    map[rawName] = id;
+
+  // Store ids as hex.
+  return Object.fromEntries(
+      Object.entries(map).map(([key, value]) => [key, parseInt(value).toString(16).toUpperCase()]),
```
Comment (quisquous) [inline util/gen_effect_id.ts:115]:
```suggestion
      Object.entries(map).map(([name, id]) => [name, parseInt(id).toString(16).toUpperCase()]),
```

Suggesting to give these better names.

File: util/gen_effect_id.ts [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+  // Make sure everything specified in known_mapping was found in the above loop.
+  for (const entry of Object.entries(knownMapping)) {
+    const name = cleanName(entry[0]);
+    if (name && !(name in foundNames))
+      printError('missing', name, knownMapping, entry[0]);
+  }
+
+  // Add custom effect name for necessary duplicates.
+  for (const [rawName, id] of Object.entries(customMapping))
+    map[rawName] = id;
```
Comment (quisquous) [inline util/gen_effect_id.ts:111]:
```suggestion
  for (const [name, id] of Object.entries(customMapping))
    map[name] = id;
```

I know the previous code said this, but I think these are technically `name` and not `rawName` (that still need `cleanName` called on them).

File: util/gen_effect_id.ts [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+const makeEffectMap = (table: Table<'#', 'Name'>) => {
+  const foundNames = new Set();
+
+  const map: { [s: string]: string } = {};
+  for (const [id, effect] of Object.entries(table)) {
+    const rawName = effect['Name'];
+    if (!rawName)
+      continue;
+    const name = cleanName(rawName);
```
Comment (quisquous) [inline util/gen_effect_id.ts:78]:
```suggestion
    const name = cleanName(rawName);
    // Skip empty strings.
    if (!name)
      continue;
```

File: util/gen_effect_id.ts (hunk 2 above)
Comment (quisquous) [reply util/gen_effect_id.ts:106]: This script now prints missing for everything in `knownMapping` even though they exist in the table.  This is a change from the previous version.

File: util/gen_effect_id.ts [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    if (rawName && rawName in knownMapping) {
+      if (id !== knownMapping[rawName as keyof typeof knownMapping]) {
+        printError('skipping', rawName, table, id);
+        continue;
+      }
+    }
+
+    if (name in map) {
+      printError('collision', name, table, id);
+      printError('collision', name, table, map[name] ?? '');
```
Comment (quisquous) [inline util/gen_effect_id.ts:89]: I think there is a bug with the collision code here, for example `Doom`:
```
collision Doom: {"#":"910","Name":"Doom","Icon":"15503","PartyListPriority":"200"}
collision Doom: {"#":"210","Name":"Doom","Icon":"15503","PartyListPriority":"200"}
collision Doom: {"#":"1769","Name":"Doom","Icon":"15503","PartyListPriority":"200"}
collision Doom: {"#":"1738","Name":"Doom","Icon":"13516","PartyListPriority":"50"}
collision Doom: {"#":"2516","Name":"Doom","Icon":"15503","PartyListPriority":"200"}
collision Doom: {"#":"1970","Name":"Doom","Icon":"15503","PartyListPriority":"200"}
```

But yet when running this script `Doom` appears in the output (where it didn't before).  It should be removed if it has a collision.

**Review by quisquous — COMMENTED** [review]
Can you delete the old script too in this PR?

File: util/gen_effect_id.ts (hunk 2 above)
Comment (MaikoTan) [reply util/gen_effect_id.ts]: because `foundNames` is a `Set`, should use `foundNames.has(name)`

File: util/gen_effect_id.ts (hunk 6 above)
Comment (MaikoTan) [reply util/gen_effect_id.ts]: Transform `map` to a truthy `Map` fixed it.

File: util/gen_effect_id.ts [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    if (name && !foundNames.has(name))
+      printError('missing', name, knownMapping, rawName);
+  }
+
+  // Add custom effect name for necessary duplicates.
+  for (const [name, id] of Object.entries(customMapping))
+    map.set(name, id);
+
+  // Store ids as hex.
+  map.forEach((value, key, map) => map.set(key, parseInt(value).toString(16).toUpperCase()));
```
Comment (quisquous) [inline util/gen_effect_id.ts:118]:
```suggestion
  map.forEach((id, name) => map.set(name, parseInt(id).toString(16).toUpperCase()));
```

**Pull request merged**